diff -ur twolame-0.3.13.orig/libtwolame/twolame.h twolame-0.3.13/libtwolame/twolame.h
--- twolame-0.3.13.orig/libtwolame/twolame.h	2011-01-02 04:15:01 +0900
+++ twolame-0.3.13/libtwolame/twolame.h	2013-04-26 20:39:20 +0900
@@ -39,7 +39,7 @@
  * the library, define LIBTWOLAME_STATIC prior to including this header.
  */
 
-#ifdef _WIN32
+#if defined(_WIN32) && !defined(__MINGW32__)
 #ifdef LIBTWOLAME_STATIC
 #define DLL_EXPORT
 #else
