--- a/loader/CMakeLists.txt
+++ b/loader/CMakeLists.txt
@@ -387,29 +387,29 @@
     endif()
 
     add_library(vulkan
-                SHARED
+                STATIC
                 ${NORMAL_LOADER_SRCS}
                 ${CMAKE_CURRENT_SOURCE_DIR}/${API_TYPE}-1.def
                 ${RC_FILE_LOCATION})
 
-    target_link_libraries(vulkan PRIVATE loader_specific_options)
+    target_link_libraries(vulkan PRIVATE $<BUILD_INTERFACE:loader_specific_options>)
 
     # when adding the suffix the import and runtime library names must be consistent
     # mingw: libvulkan-1.dll.a / vulkan-1.dll
     # msvc: vulkan-1.lib / vulkan-1.dll
-    set_target_properties(vulkan
-                          PROPERTIES
-                          OUTPUT_NAME ${API_TYPE}-1)
-    if(MINGW)
-        # generate the same DLL with mingw
-        set_target_properties(vulkan
-                              PROPERTIES
-                              PREFIX "")
-        # let the linker just fix up the stdcall mangling, like on msvc
-        if(CMAKE_SIZEOF_VOID_P EQUAL 4)
-            target_link_options(vulkan PRIVATE "-Wl,--enable-stdcall-fixup")
-        endif()
-    endif()
+    #set_target_properties(vulkan
+    #                      PROPERTIES
+    #                      OUTPUT_NAME ${API_TYPE}-1)
+    #if(MINGW)
+    #    # generate the same DLL with mingw
+    #    set_target_properties(vulkan
+    #                          PROPERTIES
+    #                          PREFIX "")
+    #    # let the linker just fix up the stdcall mangling, like on msvc
+    #    if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+    #        target_link_options(vulkan PRIVATE "-Wl,--enable-stdcall-fixup")
+    #    endif()
+    #endif()
 
    if(MSVC AND ENABLE_WIN10_ONECORE)
        target_link_libraries(vulkan PRIVATE OneCoreUAP.lib LIBCMT.LIB LIBCMTD.LIB LIBVCRUNTIME.LIB LIBUCRT.LIB)
@@ -426,3 +426,3 @@
    else()
-        add_library(vulkan SHARED)
+        add_library(vulkan STATIC)
    endif()
@@ -524,7 +524,7 @@
 endif()

 # common attributes of the vulkan library
-target_link_libraries(vulkan PRIVATE loader_specific_options)
+target_link_libraries(vulkan PRIVATE $<BUILD_INTERFACE:loader_specific_options>)
 
 target_link_libraries(vulkan PRIVATE Vulkan::Headers)
 add_library(Vulkan::Loader ALIAS vulkan)
@@ -561,14 +561,27 @@
 # https://gitlab.kitware.com/cmake/cmake/-/issues/22621
 find_package(PkgConfig)
 if (PKG_CONFIG_FOUND)
-    if(WIN32)
-        if(MINGW)
-            set(VULKAN_LIB_SUFFIX "-1.dll")
-        else()
-            set(VULKAN_LIB_SUFFIX "-1")
+    set(PRIVATE_LIBS "")
+    get_target_property(vulkan_privates vulkan LINK_LIBRARIES) # get the list of private libraries
+    foreach(LIB IN LISTS vulkan_privates)
+        if (NOT LIB MATCHES "<" AND NOT LIB MATCHES "::") # Don't include generator expressions and targets
+            list(APPEND PRIVATE_LIBS "-l${LIB}")
         endif()
+    endforeach()
+    list(REMOVE_DUPLICATES PRIVATE_LIBS)
+    string(REPLACE ";" " " PRIVATE_LIBS "${PRIVATE_LIBS}")
+    if(PRIVATE_LIBS)
+        message(NOTICE "${PRIVATE_LIBS}")
     endif()
 
+    #if(WIN32)
+    #    if(MINGW)
+    #        set(VULKAN_LIB_SUFFIX "-1.dll")
+    #    else()
+    #        set(VULKAN_LIB_SUFFIX "-1")
+    #    endif()
+    #endif()
+
     # BUG: The following code will NOT work well with `cmake --install ... --prefix <dir>`
     # due to this code relying on CMAKE_INSTALL_PREFIX being defined at configure time.
     #
--- a/loader/loader_windows.c
+++ b/loader/loader_windows.c
@@ -95,6 +95,7 @@
 #endif
 }
 
+#if 0
 BOOL WINAPI DllMain(HINSTANCE hinst, DWORD reason, LPVOID reserved) {
     (void)hinst;
     switch (reason) {
@@ -115,6 +116,8 @@
     }
     return TRUE;
 }
+#endif
+
 
 VkResult windows_add_json_entry(const struct loader_instance *inst, struct loader_string_list *search_paths,
                                 LPCSTR key_name,  // key name - used for debug prints - i.e. VulkanDriverName
--- a/loader/vulkan.pc.in
+++ b/loader/vulkan.pc.in
@@ -6,5 +6,5 @@
 Name: Vulkan-Loader
 Description: Vulkan Loader
 Version: @VULKAN_LOADER_VERSION@
-Libs: -L${libdir} -lvulkan@VULKAN_LIB_SUFFIX@
+Libs: -L${libdir} -lvulkan@VULKAN_LIB_SUFFIX@ @PRIVATE_LIBS@
 Cflags: -I${includedir}
