diff --git a/source/common/param.cpp b/source/common/param.cpp
index 00fe84900..b08282b64 100755
--- a/source/common/param.cpp
+++ b/source/common/param.cpp
@@ -2117,7 +2117,7 @@ void x265_print_params(x265_param* param)
 char *x265_param2string(x265_param* p, int padx, int pady)
 {
     char *buf, *s;
-    size_t bufSize = 4000 + p->rc.zoneCount * 64;
+    size_t bufSize = 4000 + p->rc.zoneCount * 512;
     if (p->numaPools)
         bufSize += strlen(p->numaPools);
     if (p->masteringDisplayColorVolume)
